node_monitor/node_monitor.py

```diff
@@ -98,83 +98,64 @@ def _analyze(self) -> None:
         self.actionables = {k: v for k, v
                             in self.compromised_nodes_by_provider.items()
                             if k in subscriber_ids}
-    
 
-    def broadcast_alerts(self) -> None:
-        """Broadcast relevant alerts to the appropriate channels. Retrieves
-        subscribers, node_labels, and email_recipients from the database."""
+
+    def _make_broadcaster(self) -> Callable[[str, str, str], None]:
+        """A closure that returns a broadcast function with a local cache.
+        Allows the returned function to be run in a loop without
+        querying the database.
+        """
         subscribers = self.node_provider_db.get_subscribers_as_dict()
-        node_labels = self.node_provider_db.get_node_labels_as_dict()
         email_recipients = self.node_provider_db.get_emails_as_dict()
         slack_channels = self.node_provider_db.get_slack_channels_as_dict()
         telegram_chats = self.node_provider_db.get_telegram_chats_as_dict()
-        for node_provider_id, nodes in self.actionables.items():
+
+        def broadcaster(node_provider_id: str,
+                      subject: str, message: str) -> None:
+            """Broadcasts a generic message to a subscriber through their
+            selected communication channel(s)."""
             preferences = subscribers[node_provider_id]
-            subject, message = messages.nodes_compromised_message(nodes, node_labels)
-            # - - - - - - - - - - - - - - - - -
             if preferences['notify_email'] == True:
                 recipients = email_recipients[node_provider_id]
-                logging.info(f"Sending alert email message to {recipients}...")
                 self.email_bot.send_emails(recipients, subject, message)
-            if preferences['notify_slack'] == True: 
+            if preferences['notify_slack'] == True:
                 if self.slack_bot is not None:
                     channels = slack_channels[node_provider_id]
-                    logging.info(f"Sending alert slack messages to {channels}...")
                     err1 = self.slack_bot.send_messages(channels, message)
-                    if err1 is not None:
-                        logging.error(f"SlackBot.send_message() failed with error: {err1}")
             if preferences['notify_telegram'] == True:
                 if self.telegram_bot is not None:
                     chats = telegram_chats[node_provider_id]
-                    logging.info(f"Sending alert telegram messages to {chats}...")
                     err2 = self.telegram_bot.send_messages(chats, message)
```

Comment:
are err1 and err2 being used? If not we can just remove this assignment right?

Also, shouldn't there be some kind of logging so that we know when a message was sent out on a respective channel?

Reply:
Yeah, I chose to leave these unused but still present to accommodate any changes we may want to make in the future. The values just get garbage collected.

I chose just to send one log message to not pollute our log. If you think we should add one for each channel, I can add that